chaincode/academy/go/certificate.go

```diff
@@ -2212,6 +2215,65 @@ func GetHistoryOfCertificate(stub shim.ChaincodeStubInterface, args []string) sc
 	return shim.Success(buffer.Bytes())
 }
 
+func GetSubjectsNotInCourse(stub shim.ChaincodeStubInterface, args []string) sc.Response {
+
+	if len(args) != 1 {
+		return shim.Error("Incorrect number of arguments. Expecting 1")
+	}
+
+	CourseID := args[0]
+	keyCourse := "Course-" + CourseID
+	course, err := getCourse(stub, keyCourse)
+	if err != nil {
+		return shim.Error("This course does not exist!")
+	}
+
+	allSubjects, err := getListSubjects(stub)
+	if err != nil {
+		return shim.Error("Can not get subjects list!")
+	}
+
+	defer allSubjects.Close()
+
+	var result []Subject
+	var i int
+	var subjectsInCourse = course.Subjects
+	sort.Strings(subjectsInCourse)
+	var lenSubjects = len(subjectsInCourse)
+
+	for i = 0; allSubjects.HasNext(); i++ {
+
+		record, err := allSubjects.Next()
+
+		if err != nil {
+			return shim.Success(nil)
+		}
+
+		subject := Subject{}
+		json.Unmarshal(record.Value, &subject)
+
+		if subject.SubjectID > subjectsInCourse[lenSubjects-1] || subject.SubjectID < subjectsInCourse[0] {
```

Comment:
cái `subjectInCourse` nó là một list ID, có liên tục đâu mà e lại xét kiểu này được ? 
giả sử list đấy là `[3, 7, 9]` chẳng hạn thì e bỏ sót hết các subject 4, 5, 6, 8 rồi ?

Reply:
em sắp xếp mảng subjectsInCourse từ bé đến lớn ở trên rồi ạ, nếu một subject nào đó có Id > max hoặc < min của cái mảng subjectInCourse thì chắc chắc là nó ko thuộc subjetcsInCoures, nên mình add luôn vào result và continue còn nếu không thì mình mới chạy vòng for ở dưới để kiểm tra. Như ví dụ của anh mảng subjectsInCourse đã sắp xếp là [3, 7, 9], nếu một subject có ID là 10 hoặc 2 thì mình chắc chắn là nó ko thuộc subjectsICourse nên mình add luôn vào result, còn nếu subject có Id là 4 mình chưa chắc chắn được nên sẽ chạy vòng for ở dưới ạ